File: spring-backend/src/main/java/com/simmortal/verification/VerificationService.java
package com.simmortal.verification;

import com.twilio.Twilio;
import com.twilio.rest.verify.v2.service.Verification;
import com.twilio.rest.verify.v2.service.VerificationCheck;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

@Service
public class VerificationService {
  private static final Logger logger = LoggerFactory.getLogger(VerificationService.class);
  private final String accountSid;
  private final String authToken;
  private final String serviceSid;

  public VerificationService() {
    this.accountSid = System.getenv("TWILIO_ACCOUNT_SID");
    this.authToken = System.getenv("TWILIO_AUTH_TOKEN");
    this.serviceSid = System.getenv("TWILIO_SERVICE_SID");
    if (accountSid != null && authToken != null) {
      Twilio.init(accountSid, authToken);
    }
  }

  public void sendSmsVerificationOtp(String phoneNumber) {
    ensureConfigured();
    Verification.creator(serviceSid, phoneNumber, "sms").create();
    logger.info("Sent verification SMS to {}", phoneNumber);
  }

  public boolean validateSmsVerificationOtp(String phoneNumber, String code) {
    ensureConfigured();
    VerificationCheck check = VerificationCheck.creator(serviceSid, code).setTo(phoneNumber).create();
    return "approved".equalsIgnoreCase(check.getStatus());
  }

  private void ensureConfigured() {
    if (accountSid == null || authToken == null || serviceSid == null) {
      throw new IllegalStateException("Twilio environment variables are not configured.");
    }
  }
}

File: Backend-main/src/verification/verification.service.ts
import { Injectable } from '@nestjs/common';
import { env } from 'src/config/env';
import * as createTwilioClient from 'twilio';

@Injectable()
export class VerificationService {
  private twilioClient = createTwilioClient(env.twilio.accountSid, env.twilio.authToken);

  constructor() {}

  async sendSmsVerificationOtp(phoneNumber: string): Promise<void> {
    await this.twilioClient.verify.v2
      .services(env.twilio.serviceSid)
      .verifications.create({ to: phoneNumber, channel: 'sms' });
  }

  async validateSmsVerificationOtp(phoneNumber: string, code: string): Promise<boolean> {
    const verificationCheck = await this.twilioClient.verify.v2
      .services(env.twilio.serviceSid)
      .verificationChecks.create({ to: phoneNumber, code });

    return verificationCheck.status === 'approved';
  }
}
